sources/api/netdog/src/networkd/config/netdev.rs

```diff
@@ -144,4 +147,175 @@ impl NetDevConfig {
 
         Ok(path)
     }
+
+    // The following *mut() methods are private and primarily meant for use by the NetDevBuilder.
+    // They are convenience methods to access the referenced structs (which are `Option`s) since
+    // they may need to be accessed in multiple places during the builder's construction process.
+    // (And no one wants to call `get_or_insert_with()` everywhere)
+    fn vlan_mut(&mut self) -> &mut VlanSection {
+        self.vlan.get_or_insert_with(VlanSection::default)
+    }
+
+    fn bond_mut(&mut self) -> &mut BondSection {
+        self.bond.get_or_insert_with(BondSection::default)
+    }
+}
+
+// =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+//
+/// The builder for `NetDevConfig`.
+//
+// Why a builder?  Great question.  As you can see below, some logic is involved to translate
+// config struct fields to a valid NetDevConfig.  Since `NetDevConfig` will be created by multiple
+// devices (bonds and VLANs to start), it makes sense to centralize that logic to avoid
+// duplication/mistakes.  Using a builder means type parameters can be used to limit available
+// methods based on the device being created.  Putting the type parameter on the builder and not
+// NetDevConfig avoids proliferating the type parameter everywhere NetDevConfig may be used.
+#[derive(Debug)]
+pub(crate) struct NetDevBuilder<T: Device> {
+    netdev: NetDevConfig,
+    spooky: PhantomData<T>,
+}
+
+impl<T: Device> NetDevBuilder<T> {
+    // The builder stores and directly manipulates the NetDevConfig being created to avoid storing
+    // yet another copy of the data supplied to the builder.  Generally builders have a `build()`
+    // method that returns the type being built; stick with that pattern here.  This also avoids
+    // proliferating the type parameters and provides the opportunity to add any additional
+    // creation logic needed in the future.
+    pub(crate) fn build(self) -> NetDevConfig {
+        self.netdev
+    }
+}
+
+impl NetDevBuilder<Bond> {
+    /// Create a new .netdev config for a bond.
+    pub(crate) fn new_bond(name: InterfaceName) -> Self {
+        let netdev = NetDevConfig {
+            netdev: Some(NetDevSection {
+                name: Some(name),
+                kind: Some(NetDevKind::Bond),
+            }),
+            ..Default::default()
+        };
```

Comment:
I tried to think of a way you could have a single `new` function and somehow specialize on type, but that is not possible. Overall the idea of specializing on type in order to expose only the correct functions seems consistent with Rust idioms.